module/zfs/dmu_recv.c

```diff
@@ -2043,6 +2275,60 @@ receive_process_write_record(struct receive_writer_arg *rwa,
 	    !DMU_OT_IS_VALID(drrw->drr_type))
 		return (SET_ERROR(EINVAL));
 
+	if (rwa->heal) {
+		blkptr_t *bp;
+		dmu_buf_t *dbp;
+		dnode_t *dn;
+		uint8_t buf[1];
+		int flags = DMU_READ_PREFETCH;
+
+		if (rwa->raw)
+			flags |= DMU_READ_NO_DECRYPT;
+
+		if (rwa->byteswap) {
+			dmu_object_byteswap_t byteswap =
+			    DMU_OT_BYTESWAP(drrw->drr_type);
+			dmu_ot_byteswap[byteswap].ob_func(abd_to_buf(rrd->abd),
+			    DRR_WRITE_PAYLOAD_SIZE(drrw));
+		}
+
+		/* Try to read the object to see if it needs healing */
+		err = dmu_read(rwa->os, drrw->drr_object, drrw->drr_offset,
+		    sizeof (buf), buf, flags);
+
+		/*
+		 * We only try to heal when dmu_read() returns a ECKSUMs.
+		 * Other errors (even EIO) get returned to caller.
+		 * EIO indicates that the device is not present/accessible,
+		 * so writing to it will likely fail.
+		 * If the block is healthy, we don't want to overwrite it
+		 * unnecessarily.
+		 */
+		if (err != ECKSUM)
+			return (err);
+		err = dmu_buf_hold_noread(rwa->os, drrw->drr_object,
```

Comment:
Could the dmu_read and dmu_buf_hold_noread just be replaced by a single dmu_buf_hold (since it tries the dmu_buf_hold_noread) and then test for err != ECKSUM, then proceed on using the dbp you filled in?

Reply:
Can't use `dmu_buf_hold()` since it zeros out `dbp` when we get an error (`ECKSUM` for example). However due looking at this some more I've update the code to do the `dmu_read()` after the `dmu_hold_noread()` which is what `dmu_buf_hold()` does.